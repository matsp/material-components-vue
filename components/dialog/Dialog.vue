<template>
  <div
    :aria-describedby="ariaDescribedby"
    :aria-labelledby="ariaLabelledby"
    aria-modal="true"
    class="mdc-dialog"
    :class="classes"
    role="alertdialog"
    @MDCDialog:opening="onOpening"
    @MDCDialog:opened="onOpened"
    @MDCDialog:closing="onClosing"
    @MDCDialog:closed="onClosed"
  >
    <div class="mdc-dialog__container">
      <div class="mdc-dialog__surface">
        <h2
          v-if="$slots['header']"
          :id="ariaLabelledby"
          class="mdc-dialog__title"
        >
          <slot name="header" />
        </h2>
        <div
          v-if="$slots['body']"
          :id="ariaDescribedby"
          class="mdc-dialog__content"
        >
          <slot name="body" />
        </div>
        <slot />
        <footer
          v-if="$slots['acceptButton'] || $slots['cancelButton'] || $slots['dialogButton']"
          class="mdc-dialog__actions"
        >
          <slot name="cancelButton" />
          <slot name="acceptButton" />
          <slot name="dialogButton" />
        </footer>
      </div>
    </div>
    <div class="mdc-dialog__scrim" />
  </div>
</template>

<script>
import { MDCDialog } from '@material/dialog'

import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  model: {
    prop: 'open',
    event: 'change'
  },
  props: {
    scrollable: {
      type: Boolean,
      default: false
    },
    open: {
      type: Boolean,
      default: false
    },
    stacked: {
      type: Boolean,
      default: true
    },
    escapeKeyAction: {
      type: String,
      default: 'close'
    },
    scrimClickAction: {
      type: String,
      default: 'close'
    },
    ariaDescribedby: {
      type: String,
      default: 'my-dialog-content'
    },
    ariaLabelledby: {
      type: String,
      default: 'my-dialog-title'
    }
  },
  data () {
    return {
      mdcDialog: undefined,
      slotObserver: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-dialog--scrollable': this.scrollable
      }
    },
    model: {
      get () {
        return this.open
      },
      set (value) {
        this.$emit('change', value)
      }
    }
  },
  watch: {
    open () {
      if (this.open) this.mdcDialog.open()
    },
    stacked () {
      this.mdcDialog.autoStackButtons = this.stacked
    },
    escapeKeyAction () {
      this.mdcDialog.escapeKeyAction = this.escapeKeyAction
    },
    scrimClickAction () {
      this.mdcDialog.scrimClickAction = this.scrimClickAction
    }

  },
  mounted () {
    this.mdcDialog = MDCDialog.attachTo(this.$el)
    this.mdcDialog.autoStackButtons = this.stacked
    this.mdcDialog.escapeKeyAction = this.escapeKeyAction
    this.mdcDialog.scrimClickAction = this.scrimClickAction

    this.slotObserver = new MutationObserver(() => this.updateSlots())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
    this.updateSlots()
  },
  beforeDestroy () {
    this.slotObserver.disconnect()
    this.mdcDialog.destroy()
  },
  methods: {
    onOpening () {
      this.$emit('opening')
    },
    onOpened () {
      this.$emit('opened')
      this.$el.setAttribute('aria-hidden', 'true')
    },
    onClosing (event) {
      this.$emit('closing', event.detail)
      this.$el.removeAttribute('aria-hidden')
    },
    onClosed (event) {
      this.model = false
      this.$emit('closed', event.detail)
    },
    updateSlots () {
      if (this.$slots.cancelButton) {
        this.$slots.cancelButton.map(n => {
          if (n.elm.tagName.toUpperCase() !== 'BUTTON' && !n.elm.classList.contains('mdc-dialog__button')) { n.elm.classList.add('mdc-dialog__button') }
          if (!n.elm.hasAttribute('data-mdc-dialog-action')) {
            n.elm.setAttribute('data-mdc-dialog-action', 'close')
          }
        })
      }
      if (this.$slots.acceptButton) {
        this.$slots.acceptButton.map(n => {
          if (n.elm.tagName.toUpperCase() !== 'BUTTON') {
            if (!n.elm.classList.contains('mdc-dialog__button')) {
              n.elm.classList.add('mdc-dialog__button')
            }
            if (!n.elm.classList.contains('mdc-dialog__button--default')) {
              n.elm.classList.add('mdc-dialog__button--default')
            }
          }
          if (!n.elm.hasAttribute('data-mdc-dialog-action')) {
            n.elm.setAttribute('data-mdc-dialog-action', 'accept')
          }
        })
      }
    }
  }
}
</script>
