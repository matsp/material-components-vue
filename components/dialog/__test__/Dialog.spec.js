import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import Dialog from '../Dialog.vue'

describe('Dialog', () => {
  it('should mount', () => {
    const wrapper = mount(Dialog)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(Dialog)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-dialog')
  })

  it('should render as scrollable', () => {
    const wrapper = mount(Dialog, {
      propsData: {
        scrollable: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-dialog--scrollable')
  })

  it('should render and emit', () => {
    const wrapper = mount(Dialog)

    wrapper.setProps({ open: true })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.emittedByOrder().map(e => e.name)).toEqual(['opening'])
    expect(wrapper.vm.$data.mdcDialog).toBeDefined()
    expect(wrapper.vm.$data.mdcDialog.isOpen).toBe(true)
    expect(wrapper.isVisible()).toBe(true)

    wrapper.find('.mdc-dialog__scrim').trigger('click')
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.emittedByOrder().map(e => e.name)).toEqual(['opening', 'closing'])
    expect(wrapper.emitted().closing[0]).toEqual([{ action: 'close' }])
    expect(wrapper.vm.$data.mdcDialog.isOpen).toBe(false)
  })
})
