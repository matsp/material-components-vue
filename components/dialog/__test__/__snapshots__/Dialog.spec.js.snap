// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog should render and emit 1`] = `
<div aria-describedby="my-dialog-content" aria-labelledby="my-dialog-title" aria-modal="true" role="alertdialog" class="mdc-dialog mdc-dialog--opening">
  <div class="mdc-dialog__container">
    <div class="mdc-dialog__surface">
      <!---->
      <!---->
      <!---->
    </div>
  </div>
  <div class="mdc-dialog__scrim"></div>
</div>
`;

exports[`Dialog should render and emit 2`] = `
<div aria-describedby="my-dialog-content" aria-labelledby="my-dialog-title" aria-modal="true" role="alertdialog" class="mdc-dialog mdc-dialog--opening mdc-dialog--closing">
  <div class="mdc-dialog__container">
    <div class="mdc-dialog__surface">
      <!---->
      <!---->
      <!---->
    </div>
  </div>
  <div class="mdc-dialog__scrim"></div>
</div>
`;

exports[`Dialog should render as scrollable 1`] = `
<div aria-describedby="my-dialog-content" aria-labelledby="my-dialog-title" aria-modal="true" role="alertdialog" class="mdc-dialog mdc-dialog--scrollable">
  <div class="mdc-dialog__container">
    <div class="mdc-dialog__surface">
      <!---->
      <!---->
      <!---->
    </div>
  </div>
  <div class="mdc-dialog__scrim"></div>
</div>
`;

exports[`Dialog should render with no prop 1`] = `
<div aria-describedby="my-dialog-content" aria-labelledby="my-dialog-title" aria-modal="true" role="alertdialog" class="mdc-dialog">
  <div class="mdc-dialog__container">
    <div class="mdc-dialog__surface">
      <!---->
      <!---->
      <!---->
    </div>
  </div>
  <div class="mdc-dialog__scrim"></div>
</div>
`;
